Distinguish "null" / "NULL" String values from null (Issue #6504)

catull opened the pull request:
## Impact

- [x] Bug fix (non-breaking change which fixes expected existing functionality)
 
## Description

Fixes #6504.

If a string column has the value "null" or "NULL", that column is exported as NULL when exported to SQL.

Commit:
Issue 6504 - Accept "null" / "NULL" string values.

MalloD12 commented on liquibase-standard/src/main/java/liquibase/sqlgenerator/core/InsertGenerator.java:
```diff
@@ -67,8 +68,10 @@ public void generateValues(StringBuilder sql,InsertStatement statement, Database
         for (String column : statement.getColumnValues().keySet()) {
             Object newValue = statement.getColumnValues().get(column);
-            if ((newValue == null) || "NULL".equalsIgnoreCase(newValue.toString())) {
+            if (newValue == null) {
```
Hi @catull,

I've been testing this, and I see that it does not seem to work in all cases. For example, for enumeration types, this wouldn't be valid because it will try to insert a value other than the ones valid for that given enumeration. Not sure, if there is any other type this could also affect, that I haven't realized yet.

catull replied:
Hmm, do you have a test case ?  I am not sure I can follow you.

MalloD12 replied:
Yeap, I faced this issue in the `included.changelog.xml` (which is under this path `changelogs/pgsql/complete/included.changelog.xml`). The execution of the below changeset:

```xml
 <changeSet id="6" author="molivasdat">
      <loadData file="datafiles/enum-data.csv" relativeToChangelogFile="true" tableName="sample_enum_table">
          <column header="int_col" name="int_col" type="NUMERIC"/>
          <column header="str_col" name="str_col" type="STRING"/>
          <column header="enum_col" name="enum_col" type="OTHER"/>
      </loadData>
</changeSet>
```
This was making some integration tests fail because of trying to insert a `"NULL"` into an enum type column.

catull replied:
@MalloD12

Can you be a bit more specific ?
There are about ~20 references to `included.changelog.xml`.

I'd like to investigate the test failure you refer to.

Thanks.

catull replied:
At any rate, I found the "bug".

`LoadDataChange#generateStatementsFromRows()` conflates `null` and `"NULL"`.
Both would be used as a key in a `Map`.

In order to disinguish `null` from the String `"NULL"`, I introduce a `Null` class.

Please review this, @MalloD12.

MalloD12 replied:
Hey @catull,

Are you able to see the this [execution details](https://github.com/liquibase/liquibase/actions/runs/12287878906/job/34393355812?pr=6506), you can find here what's currently failing.  If I'm not wrong, yes previously, were `PostgreSQLIntegrationTest` failing.

catull replied:
```
Error:    LoadDataOtherTest.testOtherColumnLoadData:46 expected:<NULL> but was:<null>
```

I do not get this error locally 
![image](https://github.com/user-attachments/assets/c5bfd83e-4ccb-4058-9126-7cbdef8bd3c5)

Same deal with

```
Error:    LoadDataChangeTest.NULL placeholder handling:809 Condition not satisfied:
```

Runs smoothly locally:

![image](https://github.com/user-attachments/assets/a3ea66bf-ec8d-4daa-b130-426ab1bb901a)

catull replied:
In fact, the whole test class `LoadDataChangeTest` runs all tests successfully.

catull replied:
![image](https://github.com/user-attachments/assets/b03647bc-71e5-4419-ac1b-aa613fe1b969)

catull replied:
The problem with OpenJDK 8 is, I cannot even build it on macOS, never have been.

I perform a `mvn clean package`, and I always end up with this:

![image](https://github.com/user-attachments/assets/620a3c79-e5f4-403c-9322-3747690f74c4)

Even after I `rm -fr ~/.m2/repository/jakarta`, the build never succeeds, same result as just shown.
I have stopped using OpenJDK 8 with Liquibase, when I develop with it.
I can succesfully build Liquibase with OpenJDK 11 and higher.

I use the official distribution of Liquibase in production with OpenJDK 8, OpenJDK 17 and OpenJDK 21.

catull replied:
Even this does not help at all: `mvn -DskipTests -Dmaven.javadoc.skip=true clean package`, same error.

catull replied:
This statement is from `changelog.txt`:

```
- (#5810) Bump jakarta.enterprise:jakarta.enterprise.cdi-api from 4.0.1 to 4.1.0 @dependabot
```

That version is from March 2024, when the project `jakarta.enterprise.cdi-api` jumped from JDK 11 -> JDK 17.

So, which trick do you use to build Liquibase under JDK 8 ?

catull replied:
Thanks, @MalloD12 

I had a look at a full test run under Fedora.
I believe I corrected them all now.

I get no errors now with `mvn -DtrimStackTrace=false -pl '!liquibase-cdi-jakarta' test`.

MalloD12 commented on liquibase-integration-tests/src/test/groovy/liquibase/diffchangelog/DiffChangelogIntegrationTest.groovy:
```diff
@@ -274,4 +274,36 @@ COMMENT ON COLUMN $viewName.$columnName IS '$columnComment';
+
+    def "Treat 'null' / 'NULL' values not as NULLs"() {
+        given:
+        def outputChangelogFile = "test/classes/test-null.postgres.sql"
```
Let's avoid outputting a changelog file in SQL format until we fix #6396 (I have started working on that issue), but for now I would try using XML or any other format.

catull replied:
My main motivation is changelog in SQL format.  I can wait for #6396.

MalloD12 replied:
You can ignore my previous comment because I've been trying to reproduce that issue and it seems to be working just fine now.

Commit:
Issue-6504 - fix: segregate null from "NULL" as column value.

Commit:
Issues-6504 - Fix NULL sql references.

filipelautert commented:
Hey @catull , to build under java 8 and 11 we skip the jakarta module like this:

`./mvnw -P 'skip-integration-tests'  -pl '!liquibase-cdi-jakarta' clean test package`

MalloD12 commented on liquibase-standard/src/main/java/liquibase/sqlgenerator/core/InsertGenerator.java:
```diff
@@ -67,8 +68,10 @@ public void generateValues(StringBuilder sql,InsertStatement statement, Database
         for (String column : statement.getColumnValues().keySet()) {
             Object newValue = statement.getColumnValues().get(column);
-            if ((newValue == null) || "NULL".equalsIgnoreCase(newValue.toString())) {
+            if ((newValue == null) || newValue.toString() == null) {
```
If I'm not wrong, `newValue.toString() == null` could lead us to an NPE because newValue is not a `Null` instance, so if its value is `null`, the `toString()` call will cause an exception here. Shouldn't we update this line to:

```java 
if ((newValue == null) || new Null().equals(newValue)) 
```

MalloD12 replied:
I think the second part of this IF will never be true, I think it would be just if `newValue` were of `Null` type.

MalloD12 commented:
Hi @catull,

I just left you a review comment. I'm seeing a test failure for Postgres integration tests, but this is not related to your PR,  and it is a general issue I'm also facing in #6560.  

So, after resolving the current review comment we would be good to move on with this PR.

Thank you,
Daniel.

MalloD12 commented:
Exactly, I was pointing out that in my first comment that later I deleted. Because, when  `newValue == null` would lead us to that NPE you also saw when calling the `toString()` method. Now with this change, don't you think it's just enough to ask for `if (newValue == null)`, because I think the `newValue instanceof null` doesn't add you any extra validation, it would be different if we do i`f (newValue instanceof Null)` (where `Null` is the type that you created in this PR), but that won't happen because this type is not something that will come from the DB.

catull commented:
> Exactly, I was pointing out that in my first comment that later I deleted. Because, when `newValue == null` would lead us to that NPE you also saw when calling the `toString()` method. Now with this change, don't you think it's just enough to ask for `if (newValue == null)`, because I think the `newValue instanceof null` doesn't add you any extra validation, it would be different if we do i`f (newValue instanceof Null)` (where `Null` is the type that you created in this PR), but that won't happen because this type is not something that will come from the DB.

You are right with your suggestion, @MalloD12.

I split the `if` into 2 seperate branches, to better count how many `null` and how many `Null` cases do occur.
It is a clear 552 : 0 for pure `null`.

You convinced me now, I am changing the logic it ccordingly.

Thanks for your patience.

catull committed:
Issue-6504 - Distinguish NULL value from "NULL" string.

catull committed:
Merge branch 'master' into fix-issue-6504

catull committed:
Merge branch 'master' into fix-issue-6504

catull committed:
Isue-6504 - Address Code-rabbit-bot issues.

catull committed:
Isue-6504 - Address Code-rabbit-bot issues.

MalloD12 committed:
Merge branch 'fix-issue-6504' of https://github.com/catull/liquibase into catull-fix-issue-6504

MalloD12 committed:
- Update some of th generators to check by Null type instance instead of string null value.
- Update tests to use singleton call.

MalloD12 reviewed (approved):
Hi @catull,

This is looking good to me. I have made the following changes:

- I made the updated of replacing `if ((newValue == null) || "NULL".equalsIgnoreCase(newValue.toString()))` by `if ((newValue == null) || (newValue instanceof Null))` in some os the generators (InserOrUpdateGenerator. UpdateGenerator, and InsertOrUpdateGeneratorInformix).

- Update LoadDataChangeTest ->  `new Null()` calls with `Null.getInstance()`. 
 
The rest looks good! Thank you for the changes made here.

@filipelautert are you able to authorize the test execution for this PR? I can't, but I can see it builds successfully locally.

MalloD12 committed:
Update empty string comparisson for a more efficient way.

catull committed:
Merge branch 'master' into fix-issue-6504

catull commented:
Thank you very much, @MalloD12, for your continuous support!

tati-qalified committed:
Merge branch 'master' into fix-issue-6504

tati-qalified committed:
Fix performance degradation since implementing analytics (#7439)

* Modified execute() command to create an Event object only if analytics are enabled. This will avoid creating unnecessary Event objects that automatically run resource-heavy methods and impact performance.

* Create Event only if Analytics are enabled

* Fix logic errors for isLiquibaseDocker and isAwsLiquibaseDocker

---------

Co-authored-by: tatibeans <tatiana.f.ganz@gmail.com>
Co-authored-by: filipe <flautert@liquibase.org>

chadlwilson committed:
fix: reinstate ability to run without SnakeYaml/OpenCSV on Java 25 by delaying class loading (#7430)

* refactor: avoid `Change` reference to opencsv during service load

- mitigates effect of #7318 by avoiding early references to OpenCSV during `ServiceLoader` time.
- better encapsulates opencsv usage within the CsvReader

* refactor: avoid `ChangeLog` references to SnakeYaml during service load

- mitigates effect of #7318 by avoiding early references to SnakeYaml's BaseConstructor/SafeConstructor during `ServiceLoader` time by moving the parser creation into the custom type which will only be class-loaded on use.

filipelautert committed:
fix: Support reading liquibase.classpath from properties file (#7394)

Fixes #6194

The classpath property was being ignored when specified in
liquibase.properties after version 4.19.0, causing "Cannot find
database driver" errors.

Root cause: The regex pattern in ParameterUtil only matched the
short form "classpath" but not the official configuration key
"liquibase.classpath".

Changes:
- Modified ParameterUtil.getPropertyFromInputStream() to match both
  "classpath" and "liquibase.classpath" property keys
- Added comprehensive unit tests covering both property formats,
  command line parsing, and edge cases

This maintains backward compatibility while fixing the issue where
users had to use --classpath on command line as a workaround.

mkarg committed:
Fixed issue #7415: MSSQL, SQL Anywhere: Does not quote tablespace name (#7426)

* Fixed issue #7415: MSSQL, SQL Anywhere: Does not quote tablespace name

* Tests added.

---------

Co-authored-by: Daniel Mallorga <dmallorga@liquibase.com>

dependabot[bot] committed:
chore(deps): bump the build-tools group with 4 updates (#7449)

Bumps the build-tools group with 4 updates: [org.xerial:sqlite-jdbc](https://github.com/xerial/sqlite-jdbc), [org.apache.maven.plugins:maven-resources-plugin](https://github.com/apache/maven-resources-plugin), [org.apache.maven.plugins:maven-source-plugin](https://github.com/apache/maven-source-plugin) and [org.apache.maven.plugins:maven-assembly-plugin](https://github.com/apache/maven-assembly-plugin).


Updates `org.xerial:sqlite-jdbc` from 3.51.0.0 to 3.51.1.0
- [Release notes](https://github.com/xerial/sqlite-jdbc/releases)
- [Changelog](https://github.com/xerial/sqlite-jdbc/blob/master/CHANGELOG)
- [Commits](https://github.com/xerial/sqlite-jdbc/compare/3.51.0.0...3.51.1.0)

Updates `org.apache.maven.plugins:maven-resources-plugin` from 3.3.1 to 3.4.0
- [Release notes](https://github.com/apache/maven-resources-plugin/releases)
- [Commits](https://github.com/apache/maven-resources-plugin/compare/maven-resources-plugin-3.3.1...v3.4.0)

Updates `org.apache.maven.plugins:maven-source-plugin` from 3.3.1 to 3.4.0
- [Release notes](https://github.com/apache/maven-source-plugin/releases)
- [Commits](https://github.com/apache/maven-source-plugin/compare/maven-source-plugin-3.3.1...maven-source-plugin-3.4.0)

Updates `org.apache.maven.plugins:maven-assembly-plugin` from 3.7.1 to 3.8.0
- [Release notes](https://github.com/apache/maven-assembly-plugin/releases)
- [Commits](https://github.com/apache/maven-assembly-plugin/compare/maven-assembly-plugin-3.7.1...v3.8.0)

---
updated-dependencies:
- dependency-name: org.xerial:sqlite-jdbc
  dependency-version: 3.51.1.0
  dependency-type: direct:development
  update-type: version-update:semver-patch
  dependency-group: build-tools
- dependency-name: org.apache.maven.plugins:maven-resources-plugin
  dependency-version: 3.4.0
  dependency-type: direct:production
  update-type: version-update:semver-minor
  dependency-group: build-tools
- dependency-name: org.apache.maven.plugins:maven-source-plugin
  dependency-version: 3.4.0
  dependency-type: direct:production
  update-type: version-update:semver-minor
  dependency-group: build-tools
- dependency-name: org.apache.maven.plugins:maven-assembly-plugin
  dependency-version: 3.8.0
  dependency-type: direct:production
  update-type: version-update:semver-minor
  dependency-group: build-tools
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

MalloD12 committed:
Allow skipping a missing custom change (Copy of 6627) (#7390)

* fix: Allow skipping a missing custom change

Prior to this a custom change with a class name that is not on the
classpath would always fail the entire changelog.

Now `changeSet.failOnError: false` will be honoured if the cause of the
failure is the failure to instantiate the change class.

In addition, a failing precondition for the changeset with
onFail="MARK_RAN" will be honoured if the class is not on the classpath.

Fixes https://github.com/liquibase/liquibase/issues/6520

Signed-off-by: Robert Elliot <rob@lidalia.org.uk>

* Log not loadable on custom change validation

Also fixes failing test by loading the change in generateChecksum

Signed-off-by: Robert Elliot <rob@lidalia.org.uk>

* Added additional test assertions to existent tests.

* Log whenever unable to load custom change

Signed-off-by: Robert Elliot <rob@lidalia.org.uk>

* - CustomChangeWrapper null check added.
- CustomChangeWrapper validation error/warning added to keep consistency with other scenarios instead of throwing an exceptionto keep consistency with other scenarios.
- Integration tests fixed.

* Re-apply changes of login warning messages instead of throwing an exception.

* Log message updated.

---------

Signed-off-by: Robert Elliot <rob@lidalia.org.uk>
Co-authored-by: Robert Elliot <rob@lidalia.org.uk>

dependabot[bot] committed:
chore(deps): bump the github-actions group across 1 directory with 2 updates (#7453)

Bumps the github-actions group with 2 updates in the / directory: [actions/checkout](https://github.com/actions/checkout) and [codex-/return-dispatch](https://github.com/codex-/return-dispatch).


Updates `actions/checkout` from 5 to 6
- [Release notes](https://github.com/actions/checkout/releases)
- [Changelog](https://github.com/actions/checkout/blob/main/CHANGELOG.md)
- [Commits](https://github.com/actions/checkout/compare/v5...v6)

Updates `codex-/return-dispatch` from 2 to 3
- [Release notes](https://github.com/codex-/return-dispatch/releases)
- [Commits](https://github.com/codex-/return-dispatch/compare/v2...v3)

---
updated-dependencies:
- dependency-name: actions/checkout
  dependency-version: '6'
  dependency-type: direct:production
  update-type: version-update:semver-major
  dependency-group: github-actions
- dependency-name: codex-/return-dispatch
  dependency-version: '3'
  dependency-type: direct:production
  update-type: version-update:semver-major
  dependency-group: github-actions
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

MalloD12 committed:
Fix Incompatibility Between Quarkus and Liquibase   (#7417)

* Fixed quarkus incompatibility issue.

* Supress sonar warning, since it's reported as an issue but it is a false positive.

catull committed:
Merge from master

catull committed:
Merge branch 'master' into fix-issue-6504

catull committed:
Merge branch 'master' into fix-issue-6504

MalloD12 committed:
- Fix DiffChangelogIntegrationTest by fixing sql changelog file name.
- Fix HSQL sequence next value issue.

MalloD12 committed:
Merge branch 'master' into fix-issue-6504

catull committed:
Merge branch 'master' into fix-issue-6504

catull committed:
Merge branch 'master' into fix-issue-6504

catull committed:
Merge branch 'master' into fix-issue-6504

filipelautert commented:
Hello @catull , did you see this comment here? https://github.com/liquibase/liquibase/pull/6506#issuecomment-3558962080 Seems they still failing.

catull committed:
Merge branch 'master' into fix-issue-6504

catull committed:
Merge branch 'master' into fix-issue-6504

catull committed:
Merge branch 'master' into fix-issue-6504

catull committed:
Merge branch 'master' into fix-issue-6504

catull committed:
Merge branch 'master' into fix-issue-6504

catull committed:
Merge branch 'master' into fix-issue-6504

catull committed:
Merge branch 'master' into fix-issue-6504

catull committed:
Merge branch 'master' into fix-issue-6504

catull committed:
Merge branch 'master' into fix-issue-6504

catull committed:
Merge branch 'main' into fix-issue-6504

catull committed:
Merge branch 'main' into fix-issue-6504

catull committed:
Merge branch 'main' into fix-issue-6504

catull committed:
Merge branch 'main' into fix-issue-6504

catull committed:
Merge branch 'main' into fix-issue-6504

catull committed:
Merge branch 'main' into fix-issue-6504

catull committed:
Merge branch 'main' into fix-issue-6504

catull committed:
Merge branch 'main' into fix-issue-6504

catull committed:
Merge branch 'main' into fix-issue-6504

catull committed:
Merge branch 'main' into fix-issue-6504
